generic/CTensor.c

```diff
@@ -0,0 +1,261 @@
+#ifndef THC_GENERIC_FILE
+#define THC_GENERIC_FILE "generic/CTensor.c"
+#else
+
+/* everything is as the generic Storage.c, except few things (see below) */
+
+#define TH_GENERIC_FILE "generic/Tensor.c"
+#include "generic/Tensor.c"
+#undef TH_GENERIC_FILE
+
+/* now we overwrite some methods specific to CudaTensor */
+static int cutorch_Tensor_(copy)(lua_State *L)
+{
+  THCState *state = cutorch_getstate(L);
+  THCTensor *tensor = luaT_checkudata(L, 1, torch_Tensor);
+  void *src;
+  if( (src = luaT_toudata(L, 2, "torch.CudaTensor")) )
+    THCTensor_(copyCudaFloat)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaByteTensor")) )
+    THCTensor_(copyCudaByte)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaCharTensor")) )
+    THCTensor_(copyCudaChar)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaShortTensor")) )
+    THCTensor_(copyCudaShort)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaIntTensor")) )
+    THCTensor_(copyCudaInt)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaLongTensor")) )
+    THCTensor_(copyCudaLong)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CudaDoubleTensor")) )
+    THCTensor_(copyCudaDouble)(state, tensor, src);
+
+  else if( (src = luaT_toudata(L, 2, "torch.ByteTensor")) )
+    THCTensor_(copyByte)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.CharTensor")) )
+    THCTensor_(copyChar)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.ShortTensor")) )
+    THCTensor_(copyShort)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.IntTensor")) )
+    THCTensor_(copyInt)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.LongTensor")) )
+    THCTensor_(copyLong)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.FloatTensor")) )
+    THCTensor_(copyFloat)(state, tensor, src);
+  else if( (src = luaT_toudata(L, 2, "torch.DoubleTensor")) )
+    THCTensor_(copyDouble)(state, tensor, src);
+  else
+    luaL_typerror(L, 2, "torch.*Tensor");
+
+  lua_settop(L, 1);
+  return 1;
+}
+
+static int cutorch_Tensor_(copyAsyncCPU)(lua_State *L)
+{
+  THCState *state = cutorch_getstate(L);
+  THCTensor *storage = luaT_checkudata(L, 1, "torch.CudaTensor");
+  void *src;
+  if( (src = luaT_toudata(L, 2, "torch.CudaTensor")) )
+    THCTensor_(copy)(state, storage, src);
+  else if( (src = luaT_toudata(L, 2, "torch.FloatTensor")) )
+    THCTensor_(copyAsyncCPU)(state, storage, src);
+  else
+    luaL_typerror(L, 2, "torch.FloatTensor or torch.CudaTensor");
+
+  lua_settop(L, 1);
+  return 1;
+}
```

Comment:
Shouldn't this be generic, rather than specifically for CudaTensor and FloatTensor?